agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
+	_, entries, err := store.ConfigEntriesByKind(nil, structs.APIGateway, acl.WildcardEnterpriseMeta())
 	if err != nil {
 		return err
-	} else if metaGateway == nil {
-		//delete meta gateway
-		r.logger.Info("cleaning up deleted gateway object", "request", req)
-		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
-			Kind:           structs.BoundAPIGateway,
-			Name:           req.Name,
-			EnterpriseMeta: *req.Meta,
-		}); err != nil {
-			msg := "error cleaning up deleted gateway object"
-			r.logger.Error(msg, err)
-			return errors.Wrap(err, msg)
+	}
+	requests := []controller.Request{}
+	for _, entry := range entries {
+		gateway := entry.(*structs.APIGatewayConfigEntry)
+		for _, listener := range gateway.Listeners {
+			for _, certificate := range listener.TLS.Certificates {
+				if certificate.IsSame(&structs.ResourceReference{
+					Kind:           req.Kind,
+					Name:           req.Name,
+					EnterpriseMeta: *req.Meta,
+				}) {
+					requests = append(requests, controller.Request{
+						Kind: structs.APIGateway,
+						Name: gateway.Name,
+						Meta: &gateway.EnterpriseMeta,
+					})
+				}
+			}
 		}
-		return nil
 	}
+	r.controller.Enqueue(requests...)
+	return nil
+}
 
-	r.ensureBoundGateway(metaGateway)
+func (r apiGatewayReconciler) cleanupBoundGateway(ctx context.Context, req controller.Request, store *state.Store) error {
+	logger := r.logger.With("bound-gateway", req.Name, "partition", req.Meta.PartitionOrDefault(), "namespace", req.Meta.NamespaceOrDefault())
+	logger.Debug("cleaning up bound gateway")
+	defer logger.Debug("finished cleaning up bound gateway")
 
-	routes, err := r.retrieveAllRoutesFromStore()
+	routes, err := retrieveAllRoutesFromStore(store)
 	if err != nil {
 		return err
 	}
 
-	boundGateways, routeErrors := BindRoutesToGateways([]*gatewayMeta{metaGateway}, routes...)
-
-	//In this loop there should only be 1 bound gateway returned, but looping over all returned gateways
-	//to make sure nothing gets dropped and handle case where 0 gateways are returned
-	for _, boundGateway := range boundGateways {
-		// now update the gateway state
-		r.logger.Debug("persisting gateway state", "state", boundGateway)
-		if err := r.store.Update(boundGateway); err != nil {
-			msg := "error persisting state"
-			r.logger.Error(msg, "error", err)
-			return errors.Wrap(err, msg)
-		}
-
-		// then update the gateway status
-		r.logger.Debug("persisting gateway status", "gateway", metaGateway.Gateway)
-		if err := r.store.UpdateStatus(metaGateway.Gateway, err); err != nil {
-			return err
-		}
-	}
+	resource := requestToResourceRef(req)
+	resource.Kind = structs.APIGateway
 
-	// and update the route statuses
-	for route, routeError := range routeErrors {
-		configEntry := r.resourceReferenceToBoundRoute(route)
-		r.logger.Error("route binding error:", routeError)
-		if err := r.store.UpdateStatus(configEntry, routeError); err != nil {
+	for _, toUpdate := range RemoveGateway(resource, routes...) {
```

Comment:
We handle the cleanup of route statuses in the cleanup of `BoundAPIGatewayConfigEntry` objects so that this gets triggered either on a normal detection of a gateway delete or when we GC an orphaned bound object.